### PR TITLE
Push mechanism to wait for priority execution down into diagnostics calculation.

### PR DESCRIPTION
**Pull request by ToddGrun**
Push mechanism to wait for priority execution down into diagnostics calculation.

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.InProcOrRemoteHostAnalyzerRunner.cs [hunk 1]
```diff
@@ -114,7 +114,7 @@ internal class InProcOrRemoteHostAnalyzerRunner
             var builderMap = await analysisResult.ToResultBuilderMapAsync(
                 additionalPragmaSuppressionDiagnostics, documentAnalysisScope, project, version,
                 compilationWithAnalyzers.Compilation, analyzers, skippedAnalyzersInfo,
-                compilationWithAnalyzers.AnalysisOptions.ReportSuppressedDiagnostics, cancellationToken).ConfigureAwait(false);
+                compilationWithAnalyzers.AnalysisOptions.ReportSuppressedDiagnostics, _ => new ValueTask(), cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.InProcOrRemoteHostAnalyzerRunner.cs:117]:
```suggestion
                compilationWithAnalyzers.AnalysisOptions.ReportSuppressedDiagnostics, static _ => new ValueTask(), cancellationToken).ConfigureAwait(false);
```

File: src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs [hunk 2]
```diff
@@ -144,192 +152,71 @@ internal class DiagnosticComputer
...
+                if (1 == Interlocked.Increment(ref s_highPriorityComputeTaskCount))
                 {
-                    // Step 6:
-                    //  - Stop tracking the 'computeTask' and 'cancellationTokenSource' for
-                    //    completed or cancelled task. For the case where the computeTask was
-                    //    cancelled, we will create a new 'computeTask' and 'cancellationTokenSource'
-                    //    for the retry.
-                    StopTrackingPostCompute(computeTask, cancellationTokenSource, highPriority);
-                    cancellationTokenSource.Dispose();
+                    s_highPriorityComputeTaskCompletionSource = new TaskCompletionSource<bool>();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs:174]: assert that s_highPriorityComputeTaskCompletionSource  is null.

File: src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs [hunk 3]
```diff
@@ -144,192 +152,71 @@ internal class DiagnosticComputer
...
-                        // Gracefully handle this case and ignore this exception.
+                        Contract.ThrowIfNull(s_highPriorityComputeTaskCompletionSource);
+                        s_highPriorityComputeTaskCompletionSource.SetResult(true);
+                        s_highPriorityComputeTaskCompletionSource = null;
                     }
                 }
             }
 
-            static async Task WaitForHighPriorityTasksAsync(CancellationToken cancellationToken)
+            static ValueTask WaitForPriorityExecutionAsync(CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs:196]: consider not having this local funcgion and just use a lambda you doc.

File: src/Workspaces/Core/Portable/Diagnostics/Extensions.cs [hunk 4]
```diff
@@ -204,6 +209,7 @@ private static string GetAssemblyQualifiedName(Type type)
                     {
                         foreach (var group in additionalPragmaSuppressionDiagnostics.GroupBy(d => d.Location.SourceTree!))
                         {
+                            await waitForPriorityExecutionAsync(cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Diagnostics/Extensions.cs:212]: i love all the loops we can pause now.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by ToddGrun** [comment]
Sounds like this approach won't work

**Pull request closed without merging**